File: frontend/js/market_config.js
/**
 * Client mirror of GET /api/config — single source of truth for labels and symbol.
 * Populated by app.js via applyFromApi(); never guesses Binance US vs Global.
 */

(function (global) {
  "use strict";

  var state = {
    exchange: "binance",
    exchange_label: "Binance",
    instrument_type: "spot_crypto",
    instrument_label: "Spot crypto",
    instrument_category: "spot",
    margined_futures: false,
    symbol: "BTCUSDT",
    interval: "1m",
    binance_region: "com",
    rest_base_url: "",
    ws_base_url: "",
    ui_feed_label: "Binance Spot (Global)",
    ui_market_line: "Market: BTCUSDT · Binance Spot (Global)",
    market_badge: "Binance Spot (Global) • BTCUSDT",
  };

  function assertSpotContract() {
    if (state.instrument_category !== "spot" || state.margined_futures) {
      global.console.error(
        "[RAGX] Market contract violation: dashboard is Binance Spot only.",
        {
          instrument_category: state.instrument_category,
          margined_futures: state.margined_futures,
        }
      );
    }
  }

  function applyFromApi(c) {
    if (!c || typeof c !== "object") return;
    var k;
    for (k in state) {
      if (!Object.prototype.hasOwnProperty.call(state, k) || !Object.prototype.hasOwnProperty.call(c, k)) {
        continue;
      }
      var v = c[k];
      if (v === undefined || v === null) continue;
      if (typeof v === "string" && !String(v).trim()) continue;
      state[k] = v;
    }
    assertSpotContract();
  }

  function getSymbol() {
    return String(state.symbol || "BTCUSDT").toUpperCase();
  }

  function getInterval() {
    return state.interval || "1m";
  }

  function getUiFeedLabel() {
    return state.ui_feed_label || "Binance Spot";
  }

  function getMarketBadge() {
    if (state.market_badge && String(state.market_badge).trim()) {
      return String(state.market_badge);
    }
    var venue = state.binance_region === "us" ? "US" : "Global";
    return "Binance Spot (" + venue + ") • " + getSymbol();
  }

  function getMarketLine() {
    return state.ui_market_line || "Market: " + getSymbol() + " · " + getUiFeedLabel();
  }

  function getSubtitle() {
    var iv =
      global.RagxTimeframe && typeof global.RagxTimeframe.getCurrent === "function"
        ? global.RagxTimeframe.getCurrent()
        : getInterval();
    return getSymbol() + " · " + iv + " · " + getUiFeedLabel();
  }

  function getDataStatusLine() {
    return getSymbol() + " · " + getUiFeedLabel();
  }

  function chartAriaLabel() {
    return getSymbol() + " candlestick chart · " + getUiFeedLabel();
  }

  function documentTitle() {
    return "RAGX-Trader — " + getSymbol() + " · " + getUiFeedLabel();
  }

  function getState() {
    return state;
  }

  global.RagxMarket = {
    applyFromApi: applyFromApi,
    getSymbol: getSymbol,
    getInterval: getInterval,
    getUiFeedLabel: getUiFeedLabel,
    getMarketBadge: getMarketBadge,
    getMarketLine: getMarketLine,
    getSubtitle: getSubtitle,
    getDataStatusLine: getDataStatusLine,
    chartAriaLabel: chartAriaLabel,
    documentTitle: documentTitle,
    getState: getState,
  };
})(typeof window !== "undefined" ? window : globalThis);

File: backend/market_config.py
"""
Single source of truth for the dashboard market identity (RAGX-Trader).

All live prices, candles, REST history, WebSocket klines, strategy evaluation, and logs
use the same Binance **Spot** REST/WS endpoints selected in `backend.config` plus the
active symbol from `SymbolManager`.

Defaults: **Binance Global Spot**, symbol **BTCUSDT**. Override region only with env
`RAGX_BINANCE_REGION` (see `backend.config`). There is **no** futures / perp feed.

`validate_dashboard_market_or_die()` runs after `configure_binance()` so the process
refuses to start if resolved endpoints are not the known Spot pair for the region
or if URLs contain forbidden venue fragments (futures, CME, etc.).
"""

from __future__ import annotations

import logging
from typing import Any

logger = logging.getLogger(__name__)

# Canonical identifiers (API + UI)
EXCHANGE_ID = "binance"
EXCHANGE_LABEL = "Binance"
INSTRUMENT_TYPE = "spot_crypto"
INSTRUMENT_LABEL = "Spot crypto"
INSTRUMENT_CATEGORY = "spot"
MARGINED_FUTURES = False
DEFAULT_SPOT_SYMBOL = "BTCUSDT"

# If any of these appear in REST/WS base URLs, refuse startup (wrong product / venue).
_FORBIDDEN_ENDPOINT_FRAGMENTS = (
    "fapi",
    "dapi",
    "future",
    "futures",
    "delivery",
    "cmegroup",
    "interactivebrokers",
    "deribit",
    "bybit",
    "swap",
    "perp",
    "perpetual",
    "btc1!",
    "eth1!",
)


def validate_dashboard_market_or_die() -> None:
    """
    Strict guard: active `config.rest_base()` / `ws_base_root()` must match the canonical
    Binance **Spot** endpoints for `effective_region`, and must not look like futures or
    unrelated exchanges. Call immediately after `configure_binance()`.
    """
    from backend import config

    region = config.effective_region
    exp_rest, exp_ws = config.canonical_spot_endpoints(region)
    got_rest = config.rest_base().rstrip("/")
    got_ws = config.ws_base_root().rstrip("/")
    exp_rest_n = exp_rest.rstrip("/")
    exp_ws_n = exp_ws.rstrip("/")

    if got_rest != exp_rest_n:
        msg = (
            f"Market integrity: REST base {got_rest!r} does not match required "
            f"Binance Spot ({region}) {exp_rest_n!r}. Refusing startup."
        )
        logger.error(msg)
        raise RuntimeError(msg)
    if got_ws != exp_ws_n:
        msg = (
            f"Market integrity: WebSocket base {got_ws!r} does not match required "
            f"Binance Spot ({region}) {exp_ws_n!r}. Refusing startup."
        )
        logger.error(msg)
        raise RuntimeError(msg)

    combined = (got_rest + " " + got_ws).lower()
    for frag in _FORBIDDEN_ENDPOINT_FRAGMENTS:
        if frag in combined:
            msg = (
                f"Market integrity: endpoint URLs contain forbidden fragment {frag!r} "
                "(not Binance Spot). Refusing startup."
            )
            logger.error(msg)
            raise RuntimeError(msg)

    logger.info(
        "Market integrity OK: Binance Spot (%s), REST=%s WS=%s",
        region,
        got_rest,
        got_ws,
    )


def build_api_config_payload(
    *,
    binance_region: str,
    symbol: str,
    interval: str,
) -> dict[str, Any]:
    """
    Payload for GET /api/config — the browser treats this as the active market contract.

    Requires `configure_binance()` to have run so REST/WS bases match `effective_region`.
    """
    from backend import config

    sym = str(symbol or DEFAULT_SPOT_SYMBOL).strip().upper() or DEFAULT_SPOT_SYMBOL
    iv = str(interval or "1m").strip() or "1m"
    region = binance_region if binance_region in ("com", "us") else "com"

    venue = "Global" if region == "com" else "US"
    ui_feed_label = f"Binance Spot ({venue})"
    ui_market_line = f"Market: {sym} · {ui_feed_label}"
    market_badge = f"Binance Spot ({venue}) • {sym}"

    return {
        "exchange": EXCHANGE_ID,
        "exchange_label": EXCHANGE_LABEL,
        "instrument_type": INSTRUMENT_TYPE,
        "instrument_label": INSTRUMENT_LABEL,
        "instrument_category": INSTRUMENT_CATEGORY,
        "margined_futures": MARGINED_FUTURES,
        "symbol": sym,
        "interval": iv,
        "binance_region": region,
        "rest_base_url": config.rest_base(),
        "ws_base_url": config.ws_base_root(),
        "ui_feed_label": ui_feed_label,
        "ui_market_line": ui_market_line,
        "market_badge": market_badge,
    }
